File: services/ai-analysis-service/src/repository/cveRepository.js
const pool = require('../db/pool');

/** Fetch a single CVE by ID - used by the on-demand /api/analysis/cve/:cveId route. */
async function findByCveId(cveId) {
  const { rows } = await pool.query('SELECT * FROM cve WHERE cve_id = $1', [cveId]);
  return rows[0] || null;
}

/**
 * Finds CVEs that need (re-)analysis:
 *  - never analyzed (no cve_analysis row), OR
 *  - the CVE's last_modified_date is newer than its last analyzed_at, OR
 *  - the previous analysis attempt FAILED (so it gets retried automatically).
 * Ordered by CVSS v3 score descending (nulls last) so the highest-severity
 * CVEs get analyzed first when Bedrock throughput is limited.
 */
async function findCvesNeedingAnalysis(limit) {
  const { rows } = await pool.query(
    `
    SELECT c.*
    FROM cve c
    LEFT JOIN cve_analysis a ON a.cve_id = c.cve_id
    WHERE a.cve_id IS NULL
       OR a.status = 'FAILED'
       OR c.last_modified_date > a.analyzed_at
    ORDER BY c.cvss_v3_score DESC NULLS LAST, c.last_modified_date DESC
    LIMIT $1
    `,
    [limit]
  );
  return rows;
}

module.exports = { findByCveId, findCvesNeedingAnalysis };

File: services/risk-engine-service/src/main/java/com/security/riskengine/repository/CveRepository.java
package com.security.riskengine.repository;

import com.security.riskengine.entity.Cve;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;

public interface CveRepository extends JpaRepository<Cve, String> {

    /**
     * CVEs with a successful AI analysis that either have no risk_score yet,
     * a stale one, or one computed under an older scoring model version.
     * Highest CVSS first.
     */
    @Query(value = """
        SELECT c.* FROM cve c
        JOIN cve_analysis a ON a.cve_id = c.cve_id AND a.status = 'SUCCESS'
        LEFT JOIN risk_score r ON r.cve_id = c.cve_id
        WHERE r.cve_id IS NULL
           OR a.analyzed_at > r.computed_at
           OR r.scoring_model_version <> :currentModelVersion
        ORDER BY c.cvss_v3_score DESC NULLS LAST
        LIMIT :limit
        """, nativeQuery = true)
    List<Cve> findCvesNeedingScoring(@Param("currentModelVersion") String currentModelVersion, @Param("limit") int limit);
}
